action/protocol/poll/protocol.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2018 IoTeX
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package poll
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/iotexproject/iotex-election/committee"
+	"github.com/pkg/errors"
+
+	"github.com/iotexproject/iotex-core/action"
+	"github.com/iotexproject/iotex-core/action/protocol"
+	"github.com/iotexproject/iotex-core/action/protocol/account/util"
+	"github.com/iotexproject/iotex-core/action/protocol/vote/candidatesutil"
+	"github.com/iotexproject/iotex-core/pkg/log"
+	"github.com/iotexproject/iotex-core/state"
+)
+
+const (
+	// ProtocolID defines the ID of this protocol
+	ProtocolID = "poll"
+)
+
+// Protocol defines the protocol of handling votes
+type Protocol struct {
+	electionCommittee committee.Committee
+}
+
+// NewProtocol instantiates the protocol of vote
+func NewProtocol(electionCommittee committee.Committee) *Protocol {
+	return &Protocol{electionCommittee: electionCommittee}
+}
+
+// Initialize fetches the poll result for genesis block
+func (p *Protocol) Initialize(ctx context.Context, sm protocol.StateManager, height uint64) error {
+	r, err := p.electionCommittee.ResultByHeight(height)
+	if err != nil {
+		return err
+	}
+	l := state.CandidateList{}
+	for _, c := range r.Delegates() {
+		l = append(l, &state.Candidate{
+			Address:       string(c.OperatorAddress()),
+			Votes:         c.Score(),
+			RewardAddress: string(c.RewardAddress()),
+		})
+	}
+	if err := p.validate(l); err != nil {
+		return err
+	}
+	return p.setCandidates(sm, l, uint64(1))
+}
+
+// Handle handles a vote
+func (p *Protocol) Handle(ctx context.Context, act action.Action, sm protocol.StateManager) (*action.Receipt, error) {
+	raCtx, ok := protocol.GetRunActionsCtx(ctx)
+	if !ok {
+		log.S().Panic("Miss run action context")
+	}
+	r, ok := act.(*action.PutPollResult)
+	if !ok {
+		return nil, nil
+	}
+	caller, err := util.LoadOrCreateAccount(sm, raCtx.Caller.String(), big.NewInt(0))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to load or create the account of voter %s", raCtx.Caller.String())
+	}
+	util.SetNonce(r, caller)
+
+	return nil, p.setCandidates(sm, r.Candidates(), r.Height())
+}
+
+// Validate validates a vote
+func (p *Protocol) Validate(ctx context.Context, act action.Action) error {
+	ppr, ok := act.(*action.PutPollResult)
```

Comment:
We can combine validate logic into handle, and return a failure receipt if anything is wrong.